### ISSUE TITLE
Candidate filter functions

### ISSUE THREAD

**Issue by florian-huber** [matchms/matchms-backup#75]
This was first a separate comment in #55. 
@jspaaks suggested distinguishing cleaning/filtering steps into 'pre-processing' and 'post-processing' (which I generally agree makes sense for the user as well). 

Still, maybe first a minor thing will be to solve this naming issue. We could go for:
- 'pre-processing' vs 'post-processing'
- 'cleaning' vs 'filtering'
- 'harmonization' vs 'filtering ...
Different choices might lead to slightly different splits of the functions. @jspaaks in particular asked for indicating which steps are essential for similarity score calculations (I added those to the list below).

To get an idea what we currently do to the data, here some part of our manuscript methods section:
### 1) Cleaning and completing metadata 
_Not needed for scores, but possibly for library matching._
- Make inchi strings more consistent in style (make strings consistent in style: 'INCHI="InChI=..."'). 
- Detect smiles which were given as inchi and convert those into proper inchi (openbabel).
- Search spectra with inchi but no smiles, or with smiles but no inchi. Convert annotations between smiles and inchi for those cases (openbabel).
- Search for InChI which cannot be read by openbabel.
Those were found to suffer from one of the following:
InChIkeys were given instead of InChI
InChI were incorrect. This is usually due to missing ‘-’ or other minor faults in the given strings.
If inchi are indeed inchikeys AND smiles are known, corrected inchi are produced by conversion of smiles (openbabel).
- For all other ‘defective’ inchi an automated search via Pubchem is performed. 
Potential candidates are first selected by matching compound name, and extended (if necessary) by matches of the compound formula. 
All those candidates are evaluated based on either the inchi or inchikey. A candidate is considered a match if name + inchikey (first 14 letters) or name + simplified inchi (first 3 parts of the inchi + removing problematic characters) are a match.
- Replace weird characters in Title

### 2) Correcting, completing, deriving key metadata (ionmode, charge, precursor-mass).
_Essential for modified cosine score and Spec2Vec._
- Add ionization mode (‘ionmode’ field) if missing and if possible based on adduct notation. 
- Add (and correct) signs in charge field to format: “X-” or “X+”.
- interpret charge field (based on given charge field and ionmode) 
- Derive precursor-mass and parent-mass from pepmass field 

## Filtering
_Essential for all similarity scores._
- Normalize peak intensities to max intensity peak = 1.
- Filter peaks based on min_mz, max_mz window (optional).
- Filter peaks based on exponential fit to intensitiy histogram (optional).
- Filter peaks based on relative intensity threshold (optional).

**Comment by jspaaks** [comment]
I'll see if I can spawn off some specific issues tomorrow

**Comment by jspaaks** [comment]
Let's first limit ourselves to 
1. [x]  ``normalize_intensities``: scale intensities to unit height #82 (PR: https://github.com/matchms/matchms/pull/83)
1. [x] ``select_by_mz``: select a range in mz, throw away other peaks #54 (PR: https://github.com/matchms/matchms/pull/87)
1. [x] ``select_by_intensity``: select a range in intensities, throw away other peaks #54 (PR: https://github.com/matchms/matchms/pull/87)
1. [x] ``select_by_relative_intensity``: select a range in intensities based on a threshold relative intensity, throw away other peaks #93
1. [ ] ``select_intensities_expfit``: select a range in intensities based on an exponential fit, throw away other peaks; takes some parameters like number of bins/binwidth, as well as ``one_hot`` switch. #55. 
1. [x] ``add_ionmode`` (already part of https://github.com/matchms/matchms/pull/73; refactored in https://github.com/matchms/matchms/pull/97) 
1. [x] ``add_charge`` (already part of https://github.com/matchms/matchms/pull/73; refactored in https://github.com/matchms/matchms/pull/97)
1. [ ] ``add_precursor_mass``: needs some input #104 
1. [ ] ``add_parent_mass``: needs some input #105 

Ideally, each should be its own PR

**Comment by cwmeijer** [comment]
Blocked by:
comment on and merge #73.
make new issues that describe the refactoring that's needed on #73.

**Comment by fdiblen** [comment]
#73 was merged so there is no issues blocking this.

**Comment by jspaaks** [comment]
postprocessing filters are decribed here https://github.com/matchms/matchms/issues/102. This issue can be closed when https://github.com/matchms/matchms/issues/75#issuecomment-610335520 are all checked.

**Comment by jspaaks** [comment]
In the [overview above](https://github.com/matchms/matchms/issues/75#issuecomment-610335520), there are 3 remaining filters, all of which have been spawned into their own issues. This issue can be closed.

**Comment by cwmeijer** [comment]
this item is redundant
